R/case-variables.R

```diff
@@ -0,0 +1,169 @@
+    cases <- lapply(cases, ensureValidCase)
+
+    if (!missing(else_case)) {
+        # assign API default values if not given
+        defaults <- list(id=NULL, name=NULL, numeric_value=NULL, missing=FALSE)
+        else_case <- modifyList(defaults, else_case)
+
+        if(!is.null(else_case$expression)) {
+            halt("else_cases should not have any conditions expression")
+        }
+        if(!is.character(else_case$name)) {
+            halt("else_cases must have a (character) name")
+        }
+        if(!is.null(else_case$id) & !is.whole(else_case$id)) {
+            halt("id must be an integer")
+        }
+        cases <- c(cases, list(else_case))
+    }
+
+    # make ids if the cases don't have ids.
+    need_ids <- vapply(cases, function (x) is.null(x$id), logical(1))
+    supplied_ids <- vapply(cases[!need_ids], vget("id"), numeric(1))
+    new_ids <- seq_along(cases) # generate too many ids
+    new_ids <- setdiff(new_ids, supplied_ids) # discard ids where there is overlap
+    cases[need_ids] <- lapply(seq_along(cases[need_ids]), function(i) {
+        cases[need_ids][[i]]$id <- new_ids[i]
+        return(cases[need_ids][[i]])
+    })
+
+    # check there are no duplicate ids and they are less than 2^15-1
+    all_ids <- vapply(cases, vget("id"), numeric(1))
+    if (anyDuplicated(all_ids) > 0) {
+        halt("there are duplicate ids provided: ", serialPaste(all_ids))
+    }
+    if (any(all_ids > 32767L)) {
+        halt("id must be less than 32,768, this might be a result of too many cases being used.")
+    }
+    if (any(all_ids < 1L)) {
+        halt("id must not be less than 1")
+    }
+
+    return(cases)
+}
+
+
+#' Make a case variable
+#'
+#' The `makeCaseVariable` function derives a variable using values from other
+#' variables. These are evaluated in the order they are supplied in the list
+#' as the `cases` argument (they proceed in an IF, ELSE IF, ELSE IF, ..., ELSE
+#' fashion); the first one that matches selects the corresponding value from
+#' the case list.
+#'
+#' There are two ways to specify cases, but you must pick only one (note these
+#' two will produce the same case variable):
+#'
+#' 1. When you just want to specify conditions, you can use named conditions:
+#' `makeCaseVaraible(case1=ds$v1 == 1, case2=ds$v2 == 2, name="new case")`
+#'
+#' 1. You can also use the `cases` argument, which is useful when you want to
+#' prespecify ids, or set numeric_values:
+#' `makeCaseVaraible(cases=list(list(expression=ds$v1 == 1, name="case1"), list(expression=ds$v2 == 2, name="case2")), name="new case")`
+#'
+#' @param ... a sequence of named expressions to use as cases as well as other
+#' properties to pass about the case variable (i.e. alias, description)
+#' @param cases a list of lists with each case condition to use each must
+#' include at least a `name` and an `expression` element.
+#' @param else_case a single list that has at least a `name` to serve as the
```

Comment:
Did you try my idea of allowing an "expression" be `"else"` as a magic value, so that you can just define "cases" and not have a special "else_case" argument?